File: sonic-approach/bench.go
// Quick and dirty sonic benchmark: parse kafka.json in a for loop, time it.
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/bytedance/sonic"
)

const runs = 100000

func main() {
	data, err := os.ReadFile("kafka.json")
	if err != nil {
		panic(err)
	}

	// warmup
	for i := 0; i < 1000; i++ {
		var v map[string]interface{}
		if err := sonic.Unmarshal(data, &v); err != nil {
			panic(err)
		}
	}

	start := time.Now()
	for i := 0; i < runs; i++ {
		var v map[string]interface{}
		if err := sonic.Unmarshal(data, &v); err != nil {
			panic(err)
		}
	}
	secs := time.Since(start).Seconds()

	perParseNs := secs * 1e9 / float64(runs)
	mbPerSec := (float64(len(data)) * float64(runs) / (1024.0 * 1024.0)) / secs
	parsesPerSec := float64(runs) / secs

	fmt.Printf("file: kafka.json (%d bytes)\n", len(data))
	fmt.Printf("runs: %d\n", runs)
	fmt.Printf("total: %.5f s\n", secs)
	fmt.Printf("per parse: %.1f ns\n", perParseNs)
	fmt.Printf("throughput: %.2f MB/s\n", mbPerSec)
	fmt.Printf("parses/sec: %.0f\n", parsesPerSec)
}

File: simdjson-approach/bench.cpp
// Quick and dirty simdjson benchmark: parse kafka.json in a for loop, time it.
#include <chrono>
#include <iostream>
#include "simdjson.h"
using namespace simdjson;

int main() {
    padded_string json = padded_string::load("kafka.json");
    ondemand::parser parser;

    const int RUNS = 100000;

    // warmup
    for (int i = 0; i < 1000; i++) {
        ondemand::document doc = parser.iterate(json);
        for (auto field : doc.get_object()) (void)field;
    }

    auto start = std::chrono::steady_clock::now();
    for (int i = 0; i < RUNS; i++) {
        ondemand::document doc = parser.iterate(json);
        for (auto field : doc.get_object()) (void)field;
    }
    auto end = std::chrono::steady_clock::now();

    double secs = std::chrono::duration<double>(end - start).count();
    double per_parse_ns = secs * 1e9 / RUNS;
    double mb_per_sec = (json.size() * (double)RUNS / (1024.0 * 1024.0)) / secs;
    double parses_per_sec = RUNS / secs;

    std::cout << "file: kafka.json (" << json.size() << " bytes)\n";
    std::cout << "runs: " << RUNS << "\n";
    std::cout << "total: " << secs << " s\n";
    std::cout << "per parse: " << per_parse_ns << " ns\n";
    std::cout << "throughput: " << mb_per_sec << " MB/s\n";
    std::cout << "parses/sec: " << parses_per_sec << "\n";
    return 0;
}
